xbmc/guilib/GUITextLayout.cpp

```diff
@@ -540,75 +542,93 @@ void CGUITextLayout::WrapText(const vecText &text, float maxWidth)
   if (!m_font)
     return;
 
-  int nMaxLines = (m_maxHeight > 0 && m_font->GetLineHeight() > 0)?(int)ceilf(m_maxHeight / m_font->GetLineHeight()):-1;
-
   m_lines.clear();
 
+  size_t nMaxLines;
+  if (m_maxHeight > 0 && m_font->GetLineHeight() > 0)
+  {
+    nMaxLines = static_cast<size_t>(std::ceil(m_maxHeight / m_font->GetLineHeight()));
+    if (nMaxLines == 0)
+      return;
+  }
+  else
+    nMaxLines = std::numeric_limits<size_t>::max();
+
   std::vector<CGUIString> lines;
   LineBreakText(text, lines);
 
-  for (unsigned int i = 0; i < lines.size(); i++)
+  // Split lines that exceed the maximum width,
+  // lines can be split by last space char or if there are no spaces by character
+  // to mimics the line behavior of word processors
+  for (const CGUIString& line : lines)
   {
-    const CGUIString &line = lines[i];
-    vecText::const_iterator lastSpace = line.m_text.begin();
+    if (m_lines.size() >= nMaxLines)
+      return;
+
+    if (line.m_text.empty()) // Blank line
+    {
+      m_lines.emplace_back(line);
+      continue;
+    }
+
     vecText::const_iterator pos = line.m_text.begin();
-    unsigned int lastSpaceInLine = 0;
+    vecText::const_iterator lastBeginPos = line.m_text.begin();
+    vecText::const_iterator lastSpacePos = line.m_text.end();
     vecText curLine;
-    while (pos != line.m_text.end())
+
+    while (pos < line.m_text.end())
     {
       // Get the current letter in the string
-      character_t letter = *pos;
-      // check for a space
-      if (CanWrapAtLetter(letter))
+      const character_t& letter = *pos;
+
+      if (CanWrapAtLetter(letter)) // Check for a space char
+        lastSpacePos = pos;
+
+      curLine.emplace_back(letter);
+
+      const float currWidth = m_font->GetTextWidth(curLine);
+
+      if (maxWidth > 0 && currWidth > maxWidth)
```

Comment:
0.0f for float constant.
nice to test for maxWidth > 0 but when <= 0, the result seems to be unlimited width, is that on purpose? if it is, could be optimized by copying lines to m_lines without bothering with any width measurements. I'm not sure I understand what current code doesfor maxWidth <= 0.

other corner case, seems to me that if maxWidth is between 0 and the width of the first character of a line, there will be an undefined condition for pos. It starts as line.m_text.begin() but would be decremented (pos -=1), placing it before the first element of the vector... maybe caught by debug runtime, but likely an eventual crash for release runtime as the loop keeps going.

Reply:
any comment about those special situations?

Reply:
oh yes i scrolled too fast and skipped this one...

> 0.0f for float constant.

if there is no specific Code Guidelines rule on docs i leave it as is

> nice to test for maxWidth > 0 but when <= 0, the result seems to be unlimited width, is that on purpose? if it is, could be optimized by copying lines to m_lines without bothering with any width measurements. I'm not sure I understand what current code doesfor maxWidth <= 0.

i have not considered that maxWidth could be a negative value, although it is strange to me that a GUI control have inverted values like rendered like a mirror?
however IMO an idea is NO copy the lines for maxWidth with negative values,
and copy all lines as is when maxWidth ==0
what do you think?

> other corner case, seems to me that if maxWidth is between 0 and the width of the first character of a line, there will be an undefined condition for pos. It starts as line.m_text.begin() but would be decremented (pos -=1), placing it before the first element of the vector... maybe caught by debug runtime, but likely an eventual crash for release runtime as the loop keeps going

nice one, practically there is not enough space/width to render the first character
so we can stop the loop without add any chars

Reply:
* Not sure about maxWidth <=0. I thought about it because height=-1 was a thing. maxWidth seems to come from the control width but I know too little about skins.
Do you (or our skinners @Hitcher @MikeSiLVO @jurialmunkey) have some idea / could experiment with current master to see what happens to controls with zero and negative width?
If they're clipped / invisible, an early exit would be OK here.

* maxWdith < 1st char width: it's not very nice but will do. I don't think anyone hits that case on purpose.
Maybe add a comment that a future nicer way could be to always return at least the first character of each line (it would be clipped appropriately at render time).

Reply:
more details:
controls that can use maxWidth == 0 currently are:
_FadeLabel and OSD debug info overlay_
so these has no line width limits

all other GUI controls set maxWidth that in theory should be always "> 0"
since maxWidth is calculated by x1, x2 of control coordinates,
to have a negative value means render a control like a mirror/reversed, somewhat like set control width to "-100" or so... 
i tried set label width to "-50" but weird behaviour seem to happens it appears that the value is mabe ignored and the control results with a width that is perhaps as long as the screen size!? I don't think it's a valid value...

for the situation with maxWidth == 0 such as fadelablel:
![immagine](https://github.com/xbmc/xbmc/assets/3257156/e7b843da-ef24-4614-92eb-cbeaf20f474b)
this is what happens if the maxwidth is < of the first char
the first char is rendered and clipped to control width
my pushed new changes on PR keep the first char despite his width is > of maxwidth 
this because the char has been added to curLine and last part of code after comment "Add the remaining text part"
add the char

if you try to replicate the similar situation with other controls such as label:
![immagine](https://github.com/xbmc/xbmc/assets/3257156/b0a2aab6-f9af-4374-99c0-4bee5de8dd11)
this control use ellipses `"..."` therefore if you have a too limited maxwidth will result to show a single "." dot
![immagine](https://github.com/xbmc/xbmc/assets/3257156/36ee9d5d-d51c-48ac-994c-1f0b4182716d)

so from use cases above, currently from what i can say there is this situation:
maxWidth == 0 ----> unlimited line width
maxWidth > 0 ----> regular limited line based on gui width
maxWidth < 0 ----> should never happens, but maybe better handle it, like avoid render text

considering this there would be a somewhat borderline and ambiguous case
so if a gui control like "label" that set maxWidth, and have a control width of 0
will result that the maxWidth == 0 (that which matches to unlimited maxwidth)
you probably shouldn't elaborate the lines here
but this limited case was already "unmanaged" on master

Reply:
Very nice detective work!
Are you saying that a control can have a different width and maxwidth?

I tried finding without success the call from OSD Debug Info? It seems to render with libass?

For the case maxwidth < first char width: I thought more about it, and returning the first char of the line then moving on to the next line (and forgetting about other chars of the line) is not consistent with the case where maxwidth allows 1 or more chars.

For example, for the same string ABC, with maxwidth < first char width, the result is:
A
but if maxwidth == first char width, the result is:
A
B
C
I think the second result would be less surprising and more consistent for maxwidth < first char width (of course each letter gets clipped)

Reply:
> I tried finding without success the call from OSD Debug Info? It seems to render with libass?

its at GUI level dont use libass, you can see on PR screenshots on upper left overlay debug text (guiwindowdebuginfo)

> Are you saying that a control can have a different width and maxwidth?

depends on how is implemented the code, from what i can see is not mandatory that maxwidth is equal to width GUI control, since
case 1: maxwidth == 0 is used for "unlimited" use case
case 2: CGUITextLayout class can be used without to be linked to a GUI control, such as guiwindowdebuginfo

then you are free to set any kind of maxwidth value

i think that maxwidth == 0 used for "unlimited" use case, could be changed and improved to be more consistent, this to avoid the ambiguous case mentioned on my previous comment above, but better on separate PR

> For the case maxwidth < first char width: I thought more about it, and returning the first char of the line then moving on to the next line (and forgetting about other chars of the line) is not consistent with the case where maxwidth allows 1 or more chars.

oh thanks, i think new changes pushed to pr are (i hope) much better